### DISCUSSION TITLE
Build cache & `writeFile` abstraction

### DISCUSSION THREAD

**Discussion by altano** [withastro/roadmap#288] [Proposal]
# The Idea

1. Integrations should stop writing files to disk directly and should go through an Astro abstraction for writing files.
1. That abstraction should provide a caching mechanism.
1. In the future we can use that abstraction to do other things, like providing users the ability to have a single familiar mechanism for redirecting those files to a directory of their choosing.

# Idea Background

I've been investigating why `@astrojs/image` caused my build time to shoot up to ~92 seconds.

The first issue is that it is transforming images serially which doesn't make good use of the resources available. I [submitted a PR](https://github.com/withastro/astro/pull/4626) to address that and got the build time down to ~38 seconds (with no compromises).

The second issue is where I hit a brick wall: sharp's processing of animated gifs is just slow. It's computationally expensive and just takes a while. I don't think there's a magical parameter we can pass that will fix this. The 14 PNGs in my project take <200ms and the animated gifs take the other 99% of the build time. My first idea: add the ability to pass transform customizations to `@astrojs/image` based on the file type. Either:
- allow me to ignore .gif files (since I've already optimized and resized them outside Astro) which would bring my image transformations back down to <1s, or
- allow me to pass more customizations to `sharp`. The gif formatting options has an `effort` parameter. In my testing, dropping `effort` down to `1` (fastest) makes the image transform drop from ~38 seconds to ~24 seconds. I did not study how this compromises image quality, though.

I don't love either of those approaches. The more I thought about it, the more I realized the problem isn't that image transforms take a long time, it's that those transforms have to be done again every time I `astro build` or `astro dev`. Image transforms are deterministic given the same inputs, so why are we converting images over and over again? I wouldn't mind doing a 90 second build if it was instant every time I built after that. So, the idea of a build cache was born.

# Cache Details

As a placeholder let's call the abstraction `Astro.writeZeFile(...)`.

Off the top of my head, the abstraction would need to take a few things:
- a cache key: the integration would be saying "given this serializable object I will always produce the same buffer, do what you want with that"
- where the file should be written (should Astro give full control or partial control of this?)
- a buffer to write
- the buffer's encoding

Another consideration: SSG vs. SSR. Perhaps in SSR we're returning files instead of writing them, but might still want the caching? Also SSR routes return a `Response` so does that even integrate with this idea at all?

Also: where is this being called? Right now some integrations like `astro-seo-images` write files in the `astro:build:done` hook which means they don't even work in dev mode. What can we do to make those integrations work better?

**Comment by altano** [comment]
PR with some of the ideas above is here: https://github.com/withastro/astro/pull/4765#issuecomment-1247535660 (it doesn't perfectly align with the brainstorming because the PR is less ambitious)

**Comment by regisphilibert** [comment]
I'm surprised this does not get more traction after a year. I believe a lot of people would greatly benefit from being able to cache  something else than images, like a request's response in the form a json file. Having to ping the API for every page for a menu or site meta is rather tedious, and the very helpful [Eleventy Fetch](https://github.com/11ty/eleventy-fetch) is limited to hosts with a file system...

So I for one would love to be able to handle caching with Astro.
